### PR TITLE
WWSympa: "quiet" parameter for "add"/"subscribe" requests weren't handled correctly

### LINKED ISSUE #1326: Sympa 6.2.67b.3 build failure
Version
-------

Sympa 6.2.67b.3 from https://github.com/sympa-community/sympa/releases/download/6.2.67b.3/sympa-6.2.67b.3.tar.gz

Installation method
-------------------

My own rpm, derived from official RHEL 7 rpm

Expected behavior
-----------------

6.2.67b.3 should build

Actual behavior
---------------

"make check" fails with:

```
Test Summary Report
-------------------
t/Request_Handler_add+del.t (Wstat: 512 Tests: 1 Failed: 0)
  Non-zero exit status: 2
  Parse errors: No plan found in TAP output
Files=24, Tests=1205, 58 wallclock secs ( 0.19 usr  0.02 sys + 48.61 cusr  5.04 csys = 53.86 CPU)
Result: FAIL
Failed 1/24 test programs. 0/1205 subtests failed.
```

It is possible that this is a stupid error on my part and I am just missing a dependency. However I haven't run into this sort of build failure before. Maybe the following is significant?

https://github.com/sympa-community/sympa/pull/1299

That is the only mention of "make check" on https://github.com/sympa-community/sympa/blob/6.2.67b.3/NEWS.md

### PR DESCRIPTION
**Pull request by ikedas**
- `quiet_subscription` parameter overrided `quiet` option for `add` request (fix #1326)
  - WWSympa: Adding subscribers, sometimes `quiet` parameter was omitted, because template didn't support it.

### CODE REVIEW

File: src/lib/Sympa/Request/Handler/subscribe.pm [hunk 1]
```diff
@@ -118,9 +118,7 @@ sub _twist {
     $user->lang($list->{'admin'}{'lang'}) unless $user->lang;
     $user->save;
 
-    ## Now send the welcome file to the user
-    $request->{quiet} = ($Conf::Conf{'quiet_subscription'} eq "on")
-        if $Conf::Conf{'quiet_subscription'} ne "optional";
```
Comment (ldidry) [inline src/lib/Sympa/Request/Handler/subscribe.pm:123]: I think we should keep that: deporting the `quiet` parameter in the templates, even if it’s hidden, allows users to bypass the `quiet_subscription` global parameter by editing the HTML code of the page.

Same thing in `src/lib/Sympa/Request/Handler/add.pm`

Comment (ikedas) [reply src/lib/Sympa/Request/Handler/subscribe.pm:123]: See the commit c732ce9 .  When `quiet_subscription` is `optional`, it needs to be handled in the template.  Thus, when it is `on`, it may also be handled in the template. In the end, I decided that the lines in Perl sources were not necessary.

Comment (ikedas) [reply src/lib/Sympa/Request/Handler/subscribe.pm:123]: Correction: See commit 93dc8b7 .

Comment (ldidry) [reply src/lib/Sympa/Request/Handler/subscribe.pm:123]: I understand that some things need to be handled in templates, but settings that should not be overriden by users should always be enforced in the backend otherwise users may bypass those settings.

Comment (ldidry) [reply src/lib/Sympa/Request/Handler/subscribe.pm:123]: I think that removing the check in the backend will also allow to bypass the setting when using mail commands 🤔

**Comment by ikedas** [comment]
I see.  I understood what `quiet_subscription` parameter aims.: By default (with value `optional`), it performs the previous action, i.e. `quiet` parameter is respected. Otherwise (either 'on' or 'off'), `quiet` parameter is overridden.

- c732ce9 :  This "fix" was misguided.  he problem was that default values of parameters were not handled appropriately.
- 93dc8b7 : Though part of changes by this commit may be useful, these would be better to be postponed for more review.

I'll submit another PR including corrected changes of c732ce9.  93dc8b7 will be put aside for a moment.

This PR will be closed.

**Comment by ikedas** [comment]
@ldidry , can you check commit 93dc8b7 to see if it contains the necessary changes?

**Comment by ldidry** [comment]
In https://github.com/sympa-community/sympa/commit/93dc8b7c36c4b194104b25d1bdfa0628c8f78da4, you remove the backend check of the configuration in `src/lib/Sympa/Request/Handler/subscribe.pm` and `src/lib/Sympa/Request/Handler/add.pm`:

```
$request->{quiet} = ($Conf::Conf{'quiet_subscription'} eq "on")
        if $Conf::Conf{'quiet_subscription'} ne "optional";
```

I think you should leave it the way it was. The other modifications are good.

**Comment by ikedas** [comment]
> In [93dc8b7](https://github.com/sympa-community/sympa/commit/93dc8b7c36c4b194104b25d1bdfa0628c8f78da4), you remove the backend check of the configuration in `src/lib/Sympa/Request/Handler/subscribe.pm` and `src/lib/Sympa/Request/Handler/add.pm`:
> 
> ```
> $request->{quiet} = ($Conf::Conf{'quiet_subscription'} eq "on")
>         if $Conf::Conf{'quiet_subscription'} ne "optional";
> ```
> 
> I think you should leave it the way it was. The other modifications are good.

Yes, 93dc8b7 was withdrawn and those lines were left.

How about changes on templates in this commit?

**Comment by ldidry** [comment]
- <https://github.com/sympa-community/sympa/commit/93dc8b7c36c4b194104b25d1bdfa0628c8f78da4#diff-50321a4faef6cfdcb87d4b291470de15db64915ba168b93091a75b9e60f4cc98R386>, I think you should keep the `IF conf.quiet_subscription == 'optional'`: as the `quiet` param is forced in the backend if it’s not optional, there’s no point to show the checkbox to the user, which make him believe he can change that behavior.
- <https://github.com/sympa-community/sympa/commit/93dc8b7c36c4b194104b25d1bdfa0628c8f78da4#diff-50321a4faef6cfdcb87d4b291470de15db64915ba168b93091a75b9e60f4cc98R395>: you should add an `IF conf.quiet_subscription == 'optional'`, same reason
- <https://github.com/sympa-community/sympa/commit/93dc8b7c36c4b194104b25d1bdfa0628c8f78da4#diff-2467e8e19b16c86abf4fe422cfeca4e1111452dd3d65ee85da73a6db45deb7afR16> and other similar situations: I see no point of adding an hidden quiet parameter if we force the param in the backend. If we need to keep the `quiet` param in the templates even when it’s handled in the backend, we can keep it, it’s not a big deal.
We could even, to have the `quiet` param even if `quiet_subscription` is set to `off`, do something like that:

```perl
[% IF conf.quiet_subscription == 'optional' %]
    <input id="quiet" type="checkbox" name="quiet"
        value="1"[% IF quiet %] checked="checked"[%END%] />
    <label for="quiet">[%|loc%]quiet[%END%]</label>
[%~ ELSE ~%]
    <input type="hidden" name="quiet"
        value="[% (conf.quiet_subscription == 'on') ? 1 : 0 %]" />
[% END ~%]
```

To summarise my opinion:
- keep the `IF conf.quiet_subscription == 'optional'`
- if you need to keep the `quiet` param in the templates even if the configuration does not allow users to choose it, make it hidden
- you may want to keep the `quiet` param in the templates even if the configuration says it won’t be quiet

**Comment by ikedas** [comment]
> * [93dc8b7#diff-50321a4faef6cfdcb87d4b291470de15db64915ba168b93091a75b9e60f4cc98R386](https://github.com/sympa-community/sympa/commit/93dc8b7c36c4b194104b25d1bdfa0628c8f78da4#diff-50321a4faef6cfdcb87d4b291470de15db64915ba168b93091a75b9e60f4cc98R386), I think you should keep the `IF conf.quiet_subscription == 'optional'`: as the `quiet` param is forced in the backend if it’s not optional, there’s no point to show the checkbox to the user, which make him believe he can change that behavior.
> * [93dc8b7#diff-50321a4faef6cfdcb87d4b291470de15db64915ba168b93091a75b9e60f4cc98R395](https://github.com/sympa-community/sympa/commit/93dc8b7c36c4b194104b25d1bdfa0628c8f78da4#diff-50321a4faef6cfdcb87d4b291470de15db64915ba168b93091a75b9e60f4cc98R395): you should add an `IF conf.quiet_subscription == 'optional'`, same reason

These are the modifications in `confirm_action.tt2`. I.e. whether `quiet` option should be `on` or `off` already has been decided at the previous steps. Thus, it is basically not necessary to make decision again. OTOH, if we dare to make the decision again, there is no reason to limit it to the `optional` case.

> * [93dc8b7#diff-2467e8e19b16c86abf4fe422cfeca4e1111452dd3d65ee85da73a6db45deb7afR16](https://github.com/sympa-community/sympa/commit/93dc8b7c36c4b194104b25d1bdfa0628c8f78da4#diff-2467e8e19b16c86abf4fe422cfeca4e1111452dd3d65ee85da73a6db45deb7afR16) and other similar situations: I see no point of adding an hidden quiet parameter if we force the param in the backend. If we need to keep the `quiet` param in the templates even when it’s handled in the backend, we can keep it, it’s not a big deal.

I agree. Hidden `quiet` parameters are not necessary at all.

>   We could even, to have the `quiet` param even if `quiet_subscription` is set to `off`, do something like that:
> 
> ```perl
> [% IF conf.quiet_subscription == 'optional' %]
>     <input id="quiet" type="checkbox" name="quiet"
>         value="1"[% IF quiet %] checked="checked"[%END%] />
>     <label for="quiet">[%|loc%]quiet[%END%]</label>
> [%~ ELSE ~%]
>     <input type="hidden" name="quiet"
>         value="[% (conf.quiet_subscription == 'on') ? 1 : 0 %]" />
> [% END ~%]
> ```
> 
> To summarise my opinion:
> 
> * keep the `IF conf.quiet_subscription == 'optional'`
> * if you need to keep the `quiet` param in the templates even if the configuration does not allow users to choose it, make it hidden
> * you may want to keep the `quiet` param in the templates even if the configuration says it won’t be quiet

**Comment by ikedas** [comment]
@ldidry , I rethought.

>   We could even, to have the `quiet` param even if `quiet_subscription` is set to `off`, do something like that:
> 
> ```perl
> [% IF conf.quiet_subscription == 'optional' %]
>     <input id="quiet" type="checkbox" name="quiet"
>         value="1"[% IF quiet %] checked="checked"[%END%] />
>     <label for="quiet">[%|loc%]quiet[%END%]</label>
> [%~ ELSE ~%]
>     <input type="hidden" name="quiet"
>         value="[% (conf.quiet_subscription == 'on') ? 1 : 0 %]" />
> [% END ~%]
> ```
> 
> To summarise my opinion:
> 
> * keep the `IF conf.quiet_subscription == 'optional'`
> * if you need to keep the `quiet` param in the templates even if the configuration does not allow users to choose it, make it hidden
> * you may want to keep the `quiet` param in the templates even if the configuration says it won’t be quiet

I understand that you are planning to do the following for each value of `quiet_subscription`:
  * `on`: `quiet` parameter is forced to be true.
  * `off`: `quiet` parameter is forced to be false.
  * `optional`: `quiet` parameter may be chosen by the users.

Therefore, there is no need to use form widgets (`<input>` tags) to handle `on` and `off` cases. Because they are always forced in `add`/`subscribe` handler.  Now my opinion is:

  * Keep the `IF conf.quiet_subscription == 'optional'`
  * No need for things like hidden tags to handle `on` and `off` cases.

``` code 
[% IF conf.quiet_subscription == 'optional' %]
    <input id="quiet" type="checkbox" name="quiet"
        value="1"[% IF quiet %] checked="checked"[%END%] />
    <label for="quiet">[%|loc%]quiet[%END%]</label>
[% END ~%]
```

**Comment by ldidry** [comment]
Yes, that was exactly that :slightly_smiling_face:
